### PR TITLE
Optimize evaluation during training

### LINKED ISSUE #202: [FEA] Optimize `torch_gather_like` operator used in top-k extraction
# 🚀 Feature request

## Motivation
- Create a vectorized version of  torch_gather_like function implemented [here](https://github.com/NVIDIA-Merlin/models/blob/c0df6db16180d696400e7fab9747e907eb394cf4/merlin/models/tf/utils/tf_utils.py#L108). This function is used to gather top-k indices from each row of one-hot vector of labels. 

## Your contribution

Maybe later check if we could come up with a more efficient implementation of `gather_torch_like` which does not using `TensorArray`, but rather uses only `tf.gather_nd()` with a 2D index (rows, cols), like the one I implemented here [here](https://github.com/NVIDIA-Merlin/NVTabular/blob/dc9c04d126b8ad842a32a3618e8f4495afbe0678/nvtabular/loader/tensorflow.py#L404). 
Not a priority for now, but maybe it might improve the metrics performance

_Originally posted by @gabrielspmoreira in https://github.com/NVIDIA-Merlin/models/pull/189#discussion_r813052963_

### LINKED ISSUE #210: [FEA] Computing Training/Eval metrics after each n steps
# 🚀 Feature request

Create a custom Callback for computing metrics after each n-steps instead of computing them for each batch.  This callback can be then provided to the `fit` method.

## Motivation

- Evaluating the RetrievalModels trained with negative sampling strategies requires comparing the positive scores against a set of implicit negative ones using ranking metrics. One required step to compute those metrics is the extraction of the top-k negatives.  This process can be time-consuming (when the catalog of items get large) and the goal of this feature is to gather positive scores and labels for n steps before calling the top-k extraction and ranking metrics calculation.

### PR DESCRIPTION
**Pull request by gabrielspmoreira**
This PR optimizes the evaluation during training each N steps. It also implements some refactoring for ItemRetrievalTask.

- [x] Making `BruteForceTopK` to support k=None, which computes scores over all items (used for example when no ranking metrics are used)
- [x] Renaming `pre_metrics` (e.g. BruteForceTopK) to `pre_eval_topk` (for both `ItemRetrievalTask` and `NextItemPredictionTask`), to make it clear it is only called during evaluation/prediction, and not when evaluating during training
- [x] Converting `tf_utils.gather_torch_like()` method into a vectorized version to retrieve the top-k labels. Closes #202
- [ ] Making train metrics to be computed each `train_metrics_steps`. If `train_metrics_steps=0` the no metrics are computed during training. Closes #210 **(In progress)**
- [ ] Refactoring metrics pre-sorting computation to avoid incorrect calculation (because pre-sorting and extracting top-k changes the number of relevant count for each row, which is not correct) **(in progress)**

### CODE REVIEW

File: merlin/models/tf/core.py [hunk 1]
```diff
@@ -2230,6 +2257,14 @@ def get_config(self):
         return {"block": tf.keras.utils.serialize_keras_object(self.block)}
 
 
+class IsFittingCallback(tf.keras.callbacks.Callback):
```
[the remark this answers is not in the document]
Comment (gabrielspmoreira) [reply merlin/models/tf/core.py]: Have refactored the callback to `MetricsComputeCallback`. It controls whether the metrics should be computed on a given call of `train_step()` or `test_step()`

**Comment by gabrielspmoreira** [comment]
Closing this PR, as it could not be rebased on top of this [PR](https://github.com/NVIDIA-Merlin/models/pull/189) for some reason.
I opened another draft [PR](https://github.com/NVIDIA-Merlin/models/pull/224/files) for this purpose, after fixing the merge issues.
